consul_exporter.go

```diff
@@ -0,0 +1,227 @@
+		),
+
+		serviceNodesHealthy: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Namespace: namespace,
+				Name:      "service_nodes_healthy",
+				Help:      "Is this service healthy on this node?",
+			},
+			[]string{"service", "node"},
+		),
+
+		client: consul_client,
+	}
+}
+
+// Describe describes all the metrics ever exported by the Consul exporter. It
+// implements prometheus.Collector.
+func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
+	ch <- e.up.Desc()
+	ch <- e.nodeCount.Desc()
+	ch <- e.serviceCount.Desc()
+	ch <- e.clusterServers.Desc()
+
+	e.serviceNodesTotal.Describe(ch)
+	e.serviceNodesHealthy.Describe(ch)
+}
+
+// Collect fetches the stats from configured Consul location and delivers them
+// as Prometheus metrics. It implements prometheus.Collector.
+func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
+	services := make(chan *api.ServiceEntry)
+
+	go e.queryClient(services)
+
+	e.mutex.Lock() // To protect metrics from concurrent collects.
+	defer e.mutex.Unlock()
+
+	// Reset metrics.
+	e.serviceNodesTotal.Reset()
+	e.serviceNodesHealthy.Reset()
+
+	e.setMetrics(services)
+
+	ch <- e.up
+	ch <- e.clusterServers
+	ch <- e.nodeCount
+	ch <- e.serviceCount
+
+	e.serviceNodesTotal.Collect(ch)
+	e.serviceNodesHealthy.Collect(ch)
+}
+
+func (e *Exporter) queryClient(services chan<- *api.ServiceEntry) {
+	defer close(services)
+
+	// How many peers are in the Consul cluster?
+	peers, err := e.client.Status().Peers()
+
+	if err != nil {
+		e.up.Set(0)
+		log.Printf("Query error is %v", err)
+		return
+	}
+
+	// We'll use peers to decide that we're up.
+	e.up.Set(1)
+	e.clusterServers.Set(float64(len(peers)))
+
+	// How many nodes are registered?
+	nodes, _, err := e.client.Catalog().Nodes(&api.QueryOptions{})
+
+	if err != nil {
+		// FIXME: How should we handle a partial failure like this?
+	} else {
+		e.nodeCount.Set(float64(len(nodes)))
+	}
+
+	// Query for the full list of services.
+	serviceNames, _, err := e.client.Catalog().Services(&api.QueryOptions{})
+	e.serviceCount.Set(float64(len(serviceNames)))
+
+	if err != nil {
+		// FIXME: How should we handle a partial failure like this?
+		return
+	}
+
+	e.serviceCount.Set(float64(len(serviceNames)))
+
+	for s := range serviceNames {
+		s_entries, _, err := e.client.Health().Service(s, "", false, &api.QueryOptions{})
+
+		if err != nil {
+			log.Printf("Failed to query service health: %v", err)
+			continue
+		}
+
+		// We should have one ServiceEntry per node, so use that for total nodes.
+		e.serviceNodesTotal.WithLabelValues(s).Set(float64(len(s_entries)))
+
+		for _, se := range s_entries {
+			services <- se
+		}
+	}
+}
+
+func (e *Exporter) setMetrics(services <-chan *api.ServiceEntry) {
+	for entry := range services {
+		// We have a Node, a Service, and one or more Checks. Our
+		// service-node combo is passing if all checks have a `status`
+		// of "passing."
+
+		passing := 1
+
+		for _, hc := range entry.Checks {
+			if hc.Status != "passing" {
```

Comment:
We should check against the constant consul provides instead of using the string here.

```go
import consul "github.com/hashicorp/consul/consul/structs"
if check.Status != consul.HealthPassing {
  ...
}
```